ws.c

```diff
@@ -0,0 +1,433 @@
+
+	if (length_field <= 125) {
+		payload_len = length_field;
+	} else if (length_field == 126) { /* msglen is 16bit */
+		if (buf_len < 4)
+			return INCOMPLETE_DATA;
+		payload_len = ntohs(*(uint16_t *)(in_buffer + 2));
+		pos += 2;
+	} else if (length_field == 127) { /* msglen is 64bit */
+		if (buf_len < 10)
+			return INCOMPLETE_DATA;
+		payload_len = ntohs(*(uint64_t *)(in_buffer + 2));
+		pos += 8;
+	}
+	if (buf_len < payload_len + pos + (masked ? 4 : 0)) {
+		return INCOMPLETE_DATA;
+	}
+
+	/* According to RFC it seems that unmasked data should be prohibited
+	 * but we support it for nonconformant clients
+	 */
+	if (masked) {
+		unsigned char *c;
+
+		mask = *((unsigned int *)(in_buffer + pos));
+		pos += 4;
+
+		/* unmask data */
+		c = in_buffer + pos;
+		for (int i = 0; i < payload_len; i++) {
+			c[i] = c[i] ^ ((unsigned char *)(&mask))[i % 4];
+		}
+	}
+
+	*payload_ptr = in_buffer + pos;
+	*out_len = payload_len;
+
+	/* are reserved for further frames */
+	if ((opcode >= 3 && opcode <= 7) || (opcode >= 0xb))
+		return ERROR_FRAME;
+
+	if (opcode <= 0x3 && !fin) {
+		return INCOMPLETE_FRAME;
+	}
+	return opcode;
+}
+
+
+static void
+ws_evhttp_read_cb(struct bufferevent *bufev, void *arg)
+{
+	struct evws_connection *evws = arg;
+	unsigned char *payload;
+	enum WebSocketFrameType type;
+	int msg_len, in_len, header_sz;
+	struct evbuffer *input = bufferevent_get_input(evws->bufev);
+
+	while ((in_len = evbuffer_get_length(input))) {
+		unsigned char *data = evbuffer_pullup(input, in_len);
+
+		type = get_ws_frame(data, in_len, &payload, &msg_len);
+		if (type == INCOMPLETE_DATA) {
+			/* incomplete data received, wait for next chunk */
+			return;
+		}
+		header_sz = payload - data;
+		evbuffer_drain(input, header_sz);
+		data = evbuffer_pullup(input, -1);
+
+		switch (type) {
+		case TEXT_FRAME:
+		case BINARY_FRAME:
+			if (evws->incomplete_frames != NULL) {
+				/* we already have incomplete frames in internal buffer
+				 * and need to concatenate them with final */
+				evbuffer_add(evws->incomplete_frames, data, msg_len);
+				evws->cb(evws, type, evws->incomplete_frames,
+					evbuffer_get_length(evws->incomplete_frames), evws->cb_arg);
+				evbuffer_free(evws->incomplete_frames);
+				evws->incomplete_frames = NULL;
+			} else {
+				evws->cb(evws, type, input, msg_len, evws->cb_arg);
+			}
+			break;
+		case INCOMPLETE_FRAME:
+			/* we received full frame until get fin and need to
+			 * postpone callback until all data arrives */
+			if (evws->incomplete_frames == NULL) {
+				evws->incomplete_frames = evbuffer_new();
+			}
+			evbuffer_remove_buffer(input, evws->incomplete_frames, msg_len);
+			continue;
+		case CLOSING_FRAME:
+		case ERROR_FRAME:
+			evws_force_disconnect_(evws);
+			break;
+		case PING_FRAME:
+		case PONG_FRAME:
+			/* ping or pong frame */
+			break;
+		default:
+			event_warn("%s: unexpected frame type %d\n", __func__, type);
+			evws_force_disconnect_(evws);
+		}
+		evbuffer_drain(input, msg_len);
```

Comment:
Should this be done in case of incomplete frames?

Reply:
Incomplete frames will not pass through here because of `continue` operator which do next iteration for the loop